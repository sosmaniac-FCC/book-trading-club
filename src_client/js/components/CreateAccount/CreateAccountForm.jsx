import React from 'react';
import { Link } from 'react-router-dom';

const CreateAccountForm = (props) => (
    <form className="col s10 offset-s1" id="create-account" onSubmit={props.handleDefaultSubmit}>
        <div id="border-sides-create">
            <div className="row">
                <div className="input-field col s10 offset-s1 m5 offset-m1">
                    <input type="text" name="username" id="username" className="validate"/>
                    <label htmlFor="username" data-success="Usernames are permanent upon creation">Username</label>
                </div>
                <div className="input-field col s10 offset-s1 m5">
        			<input type="email" name="email" id="email" className="validate"/>
        			<label htmlFor="email">Email</label>
        		</div>
    		</div>
    		
    		<div className="row">
                <div className="input-field col s10 offset-s1 m5 offset-m1">
                    <input type="text" name="city" id="city" className="validate"/>
                    <label htmlFor="city">City</label>
                </div>
                <div className="input-field col s10 offset-s1 m5">
        			<input type="text" name="state" id="state" className="validate"/>
        			<label htmlFor="state">State/Province</label>
        		</div>
    		</div>
    		
    		<div className="row">
                <div className="input-field col s10 offset-s1 m5 offset-m1">
                    <input type="password" name="password" id="password" className="validate"/>
                    <label htmlFor="password">Password</label>
                </div>
                <div className="input-field col s10 offset-s1 m5">
        			<input type="password" name="confirm" id="confirm" className="validate"/>
        			<label htmlFor="confirm">Confirm Password</label>
        		</div>
    		</div>
    	</div>
    	
    	<div className="row center">
    	    <button className="waves-effect waves-light btn purple accent-2 remove-button-padding" type="button"><Link className="white-text add-button-padding" to={props.matchUrl.substring(0, props.matchUrl.lastIndexOf('/'))}>Cancel</Link></button>
    	    {"     "}
    	    <button className="waves-effect waves-light btn purple accent-1 add-button-padding" type="submit">Create Account</button>
    	</div>
    </form>
);

export default CreateAccountForm;