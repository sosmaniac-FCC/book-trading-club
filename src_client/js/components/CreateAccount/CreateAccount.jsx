import React from 'react';
import { Redirect } from 'react-router-dom';

import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import { clearReason, registerNewUser } from '../../reducers/userReducer/userActions.jsx';
import Alert from './Alert.jsx';
import CreateAccountForm from './CreateAccountForm';
import Preloader from '../Preloader/Preloader.jsx';

/* global $ */

class CreateAccount extends React.Component {
    constructor(props) {
        super(props);
        
        this.handleDefaultSubmit = this.handleDefaultSubmit.bind(this);
        
        this.state = {
            redirectToLogin: false
        };
    }
    
    handleDefaultSubmit(e) {
        e.preventDefault();
    }
    
    componentWillMount() {
        this.props.clearReason();
    }
    
    initializeValidator() {
        $.validator.setDefaults({
            errorClass: 'invalid', // false
            validClass: "valid", // true
            errorPlacement: (error, element) => {
                $(element).closest("form").find("label[for='" + element.attr("id") + "']").attr("data-error", error.text());
            },
            submitHandler: async (form) => {
                // console.log('B');
                
                const elements = form.getElementsByTagName("input");
                let inputs = {};
                
                for (let i = 0; i < elements.length; i++) {
                    inputs[elements[i].id] = elements[i].value;
                }
                
                const isCreated = await new Promise((resolve, reject) => {
                    this.props.registerNewUser(resolve, reject, inputs);
                });
                
                if (isCreated) {
                    this.setState({
                        redirectToLogin: true
                    });
                }
            }
        });
        
        this.validator = $("#create-account").validate({
            rules: {
                username: {
                    required: true,
                    minlength: 3,
                    maxlength: 12
                },
                email: {
                    email: true,
                    required: true
                },
                city: {
                    required: true
                },
                state: {
                    required: true
                },
                password: {
                    required: true
                },
                confirm: {
                    required: true,
                    equalTo: "#password"
                }
            },
            messages: {
                username: "Username between 3 to 12 characters required",
                email: "Valid email address required",
                city: "City is required",
                state: "State/Province is required",
                password: "Password is required",
                confirm: "Passwords must match"
            }
        });
    }
    
    componentDidUpdate() {
        this.initializeValidator();
    }
    
    componentDidMount() {
        this.initializeValidator();
    }
    
    render() {
        if (this.props.loading) {
            this.skip = false;
            return <Preloader/>;
        }
        
        if (this.state.redirectToLogin) {
            return <Redirect to={{
                pathname: "/Login",
                state: "Account successfully created."
            }}/>;
        }
        
        return (
            <div className="container" id="create">
                <div className="row">
                    <h3 className="col s12 center">New Account</h3>
                    { this.props.reason ? <Alert message={this.props.reason}/> : null }
                    <CreateAccountForm handleDefaultSubmit={this.handleDefaultSubmit} matchUrl={this.props.match.url}/>
                </div>
            </div>
        );
    }
}

function mapStateToProps({ users }) {
    return {
        reason: users.reason,
        loading: users.loading
    };
}

function mapDispatchToProps(dispatch) {
    return bindActionCreators({
        clearReason,
        registerNewUser
    }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(CreateAccount);