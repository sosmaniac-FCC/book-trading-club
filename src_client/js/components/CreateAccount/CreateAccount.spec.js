// expect from jest_expect.js
/* global expect */

import React from 'react';
import { shallow, configure } from 'enzyme';
import enzymeAdapter from 'enzyme-adapter-react-16';
import toJSON from 'enzyme-to-json';

configure({ adapter: new enzymeAdapter() });

import CreateAccount from './CreateAccount.jsx';

test('Account creation page loads as expected.', () => {
    const component = shallow(<CreateAccount />);
    const tree = toJSON(component);
    expect(tree).toMatchSnapshot();
});

test('Account creation page should have no style attributes.', () => {
    const component = shallow(<CreateAccount />);
    expect(component.find('[style]')).toHaveLength(0);
});