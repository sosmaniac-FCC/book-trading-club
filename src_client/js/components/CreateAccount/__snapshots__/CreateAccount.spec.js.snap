// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account creation page loads as expected. 1`] = `
<div>
  <h1>
    Account Creation
  </h1>
</div>
`;
